File: ManOpenSwift/AproposDocument.m
/* AproposDocument.m created by lindberg on Tue 10-Oct-2000 */

#import "AproposDocument.h"
#import "ManDocumentController.h"
#import "PrefPanelController.h"

@implementation AproposDocument
@synthesize title;
@synthesize tableView;
@synthesize titleColumn;

+ (BOOL)canConcurrentlyReadDocumentsOfType:(NSString *)typeName
{
    return YES;
}

- (void)_loadWithString:(NSString *)apropos manPath:(NSString *)manPath title:(NSString *)aTitle
{
    ManDocumentController *docController = [ManDocumentController sharedDocumentController];
    NSMutableString *command = [docController manCommandWithManPath:manPath];
    NSData *output;
    
    titles = [[NSMutableArray alloc] init];
    descriptions = [[NSMutableArray alloc] init];
    title = aTitle;
    [self setFileType:@"apropos"];

    /* Searching for a blank string doesn't work anymore... use a catchall regex */
    if ([apropos length] == 0)
        apropos = @".";
    searchString = apropos;

    /*
     * Starting on Tiger, man -k doesn't quite work the same as apropos directly.
     * Use apropos then, even on Panther.  Panther/Tiger no longer accept the -M
     * argument, so don't try... we set the MANPATH environment variable, which
     * gives a warning on Panther (stderr; ignored) but not on Tiger.
     */
    // [command appendString:@" -k"];
    [command setString:@"/usr/bin/apropos"];
    
    [command appendFormat:@" %@", EscapePath(apropos, YES)];
    output = [docController dataByExecutingCommand:command manPath:manPath];
    /* The whatis database appears to not be UTF8 -- at least, UTF8 can fail, even on 10.7 */
    [self parseOutput:[[NSString alloc] initWithData:output encoding:NSMacOSRomanStringEncoding]];
}

- (id)initWithString:(NSString *)apropos manPath:(NSString *)manPath title:(NSString *)aTitle
{
    if(self = [super init]) {
        [self _loadWithString:apropos manPath:manPath title:aTitle];
        
        if ([titles count] == 0) {
            NSRunAlertPanel(@"Nothing found", @"No pages related to '%@' found", nil, nil, nil, apropos);
            return nil;
        }
    }
    return self;
}

- (NSString *)windowNibName
{
    return @"Apropos";
}

- (NSString *)displayName
{
    return title;
}

- (void)windowControllerDidLoadNib:(NSWindowController *)windowController
{
    NSString *sizeString = [[NSUserDefaults standardUserDefaults] stringForKey:@"AproposWindowSize"];

    [super windowControllerDidLoadNib:windowController];

    if (sizeString != nil)
    {
        NSSize windowSize = NSSizeFromString(sizeString);
        NSWindow *window = [tableView window];
        NSRect frame = [window frame];

        if (windowSize.width > 30.0 && windowSize.height > 30.0) {
            frame.size = windowSize;
            [window setFrame:frame display:NO];
        }
    }

    [tableView setTarget:self];
    [tableView setDoubleAction:@selector(openManPages:)];
    [tableView sizeLastColumnToFit];
}

- (void)parseOutput:(NSString *)output
{
    NSArray *lines = [output componentsSeparatedByString:@"\n"];
    if ([output length] == 0) return;

    lines = [lines sortedArrayUsingSelector:@selector(caseInsensitiveCompare:)];
    
    for (NSString *line in lines)
    {
        NSRange dashRange;

        if ([line length] == 0) continue;

        dashRange = [line rangeOfString:@"\t\t- "]; //OPENSTEP
        if (dashRange.length == 0)
            dashRange = [line rangeOfString:@"\t- "]; //OPENSTEP
        if (dashRange.length == 0)
            dashRange = [line rangeOfString:@"\t-" options:NSBackwardsSearch|NSAnchoredSearch];
        if (dashRange.length == 0)
            dashRange = [line rangeOfString:@" - "]; //MacOSX
        if (dashRange.length == 0)
            dashRange = [line rangeOfString:@" -" options:NSBackwardsSearch|NSAnchoredSearch];

        if (dashRange.length == 0) continue;

        [titles addObject:[line substringToIndex:dashRange.location]];
        [descriptions addObject:[line substringFromIndex:NSMaxRange(dashRange)]];
    }
}

- (IBAction)saveCurrentWindowSize:(id)sender
{
    NSSize size = [[tableView window] frame].size;
    [[NSUserDefaults standardUserDefaults] setObject:NSStringFromSize(size) forKey:@"AproposWindowSize"];
}

- (IBAction)openManPages:(id)sender
{
    if ([sender clickedRow] >= 0) {
        NSString *manPage = titles[[sender clickedRow]];
        [[ManDocumentController sharedDocumentController] openString:manPage oneWordOnly:YES];
    }
}

- (void)printDocumentWithSettings:(NSDictionary *)printSettings showPrintPanel:(BOOL)showPanel delegate:(id)delegate didPrintSelector:(SEL)didPrintSelector contextInfo:(void *)contextInfo;
{
    NSPrintOperation *op = [NSPrintOperation printOperationWithView:tableView];
    [op setShowsPrintPanel:showPanel];
    [op setShowsProgressPanel:showPanel];
    [op runOperationModalForWindow:[tableView window] delegate:delegate didRunSelector:didPrintSelector contextInfo:contextInfo];
}

/* NSTableView dataSource */
- (NSInteger)numberOfRowsInTableView:(NSTableView *)tableView
{
    return [titles count];
}

- (id)tableView:(NSTableView *)tableView objectValueForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row
{
    NSArray *strings = (tableColumn == titleColumn) ? titles : descriptions;
    return strings[row];
}

/* Document restoration */
#define RestoreSearchString @"SearchString"
#define RestoreTitle @"Title"

- (void)encodeRestorableStateWithCoder:(NSCoder *)coder
{
    [super encodeRestorableStateWithCoder:coder];
    [coder encodeObject:searchString forKey:RestoreSearchString];
    [coder encodeObject:title forKey:RestoreTitle];
}

- (void)restoreStateWithCoder:(NSCoder *)coder
{
    [super restoreStateWithCoder:coder];
    
    if (![coder containsValueForKey:RestoreSearchString])
        return;
    
    NSString *search = [coder decodeObjectForKey:RestoreSearchString];
    NSString *theTitle = [coder decodeObjectForKey:RestoreTitle];
    NSString *manPath = [[NSUserDefaults standardUserDefaults] manPath];
    
    [self _loadWithString:search manPath:manPath title:theTitle];
    [[self windowControllers] makeObjectsPerformSelector:@selector(synchronizeWindowTitleWithDocumentName)];
    [tableView reloadData];
}

@end

File: ManOpenSwift/AproposDocument.swift
//
//  AproposDocument.swift
//  ManOpen
//
//  Created by C.W. Betts on 9/9/14.
//
//

import Cocoa

private let restoreSearchString = "SearchString"
private let restoreTitle = "Title"

class AproposDocument: NSDocument, NSTableViewDataSource {
	var title: String = ""
	var searchString: String = ""
	var titles = [String]()
	var descriptions = [String]()
	@IBOutlet weak var tableView: NSTableView!
	@IBOutlet weak var titleColumn: NSTableColumn!

	override class func canConcurrentlyReadDocumentsOfType(typeName: String) -> Bool {
		return true
	}
	
	override var displayName: String {
		return title
	}
	
    override var windowNibName: String {
        return "Apropos"
    }

	func parseOutput(output: String) {
		
	}
	
    override func windowControllerDidLoadNib(aController: NSWindowController?) {
		var aSizeString = NSUserDefaults.standardUserDefaults().stringForKey("AproposWindowSize")

        super.windowControllerDidLoadNib(aController)
        // Add any code here that needs to be executed once the windowController has loaded the document's window.
		if let sizeString = aSizeString {
			var windowSize = NSSizeFromString(sizeString)
			var window = tableView.window
			var frame = window!.frame
			
			if windowSize.width > 30.0 && windowSize.height > 30.0 {
				frame.size = windowSize
				window!.setFrame(frame, display: false)
			}
		}
		
		tableView.target = self
		tableView.doubleAction = "openManPages:"
		tableView.sizeLastColumnToFit()
    }

    override func dataOfType(typeName: String?, error outError: NSErrorPointer) -> NSData? {
        // Insert code here to write your document to data of the specified type. If outError != NULL, ensure that you create and set an appropriate error when returning nil.
        // You can also choose to override -fileWrapperOfType:error:, -writeToURL:ofType:error:, or -writeToURL:ofType:forSaveOperation:originalContentsURL:error: instead.
		outError.memory = NSError(domain: NSOSStatusErrorDomain, code: unimpErr, userInfo: nil)
        return nil
    }

    override func readFromData(data: NSData?, ofType typeName: String?, error outError: NSErrorPointer) -> Bool {
        // Insert code here to read your document from the given data of the specified type. If outError != NULL, ensure that you create and set an appropriate error when returning NO.
        // You can also choose to override -readFromFileWrapper:ofType:error: or -readFromURL:ofType:error: instead.
        // If you override either of these, you should also override -isEntireFileLoaded to return NO if the contents are lazily loaded.
		outError.memory = NSError(domain: NSOSStatusErrorDomain, code: unimpErr, userInfo: nil)
        return false
    }

	private func loadWithString(apropos: String, manPath: String, title aTitle: String) {
		let docController = ManDocumentController.sharedDocumentController() as ManDocumentController
		var command = docController.manCommandWithManPath(manPath)
		
		title = aTitle
		self.fileType = "apropos"
	}

	override func encodeRestorableStateWithCoder(coder: NSCoder) {
		super.encodeRestorableStateWithCoder(coder)
		coder.encodeObject(searchString, forKey: restoreSearchString)
		coder.encodeObject(title, forKey: restoreTitle)
	}
	
	override func restoreStateWithCoder(coder: NSCoder) {
		super.restoreStateWithCoder(coder)
		
		if !coder.containsValueForKey(restoreSearchString) {
			return
		}
		
		var search: String = coder.decodeObjectForKey(restoreSearchString) as NSString
		var theTitle = coder.decodeObjectForKey(restoreTitle) as NSString
		var manPath = NSUserDefaults.standardUserDefaults().manPath
		
		loadWithString(search, manPath: manPath, title: theTitle)
		
		//(self.windowControllers).makeObjectsPerformSelector("synchronizeWindowTitleWithDocumentName")
		for controller in self.windowControllers as [NSWindowController] {
			controller.synchronizeWindowTitleWithDocumentName()
		}
		tableView.reloadData()
	}
	
	init?(string apropos: String, manPath: String, title: String) {
		super.init()
		return nil
	}
}
